includes/model/class-openedx-woocommerce-plugin-enrollment.php

```diff
@@ -0,0 +1,785 @@
+        }
+    }
+
+    /**
+     * Save post metadata when a post is saved.
+     *
+     * @param int  $post_id The post ID.
+     * @param bool $force Does this order need processing by force?.
+     */
+    public function process_request( $post_id, $force, $do_pre_enroll = true ) {
+
+        $user_args = $this->prepare_args( $post_id, 'user' );
+        $user      = WP_EoxCoreApi()->get_user_info( $user_args );
+
+        // If the user doesn't exist create pre-enrollment with the email provided.
+        if ( is_wp_error( $user ) && $do_pre_enroll ) {
+            if ( ! empty( $user_args['email'] ) ) {
+                $pre_enrollment_args = $this->prepare_args( $post_id, 'pre-enrollment' );
+                $this->create_pre_enrollment( $post_id, $pre_enrollment_args );
+                return;
+            } else {
+                // TODO Polish error message display.
+                update_post_meta( $post_id, 'errors', 'A valid username or email is needed.' );
+                $this->update_post_status( 'eor-error', $post_id );
+                return;
+            }
+        }
+        $enrollment_args          = $this->prepare_args( $post_id, 'enrollment' );
+        $enrollment_args['force'] = $force;
+
+        // When we reach this line we already have the user as a response from the API.
+        // Better use that username in case anything from before does not match.
+        $enrollment_args['username'] = $user->username;
+
+        $enrollment = WP_EoxCoreApi()->get_enrollment( $enrollment_args );
+
+        // If the enrollment already exists update it.
+        if ( is_wp_error( $enrollment ) ) {
+            $this->create_enrollment( $post_id, $enrollment_args );
+        } else {
+            $this->update_enrollment( $post_id, $enrollment_args );
+        }
+    }
+
+    /**
+     * Prepare args to be passed to the api calls
+     *
+     * @param int    $post_id  The post ID.
+     * @param string $type The args type to be prepared (user, enrollment, ..).
+     *
+     * @return array args args ready to be pass
+     */
+    public function prepare_args( $post_id, $type ) {
+
+        $args      = array();
+        $user_args = array_filter(
+            array(
+                'email'    => get_post_meta( $post_id, 'email', true ),
+                'username' => get_post_meta( $post_id, 'username', true ),
+            )
+        );
+
+        $enrollment_args = array(
+            'course_id' => get_post_meta( $post_id, 'course_id', true ),
+        );
+
+        $enrollment_opts_args = array(
+            'mode'      => get_post_meta( $post_id, 'mode', true ),
+            'is_active' => ( get_post_meta( $post_id, 'is_active', true ) ? true : false ),
+        );
+
+        switch ( $type ) {
+            case 'user':
+                return $user_args;
+            case 'enrollment':
+                // By the API design enrollment operations work on usernames.
```

Comment:
this will change since we are defining the API again. This time we can use the email instead. Note for our plugin @MaferMazu

Reply:
Noted!